File: old/RPINeopixel/src/Applications/RacingGame/RacingGame.cpp
//
// Created by tim on 08.12.20.
//

#include <cstring>
#include <sys/time.h>
#include "RacingGame.h"

#ifndef RACINGGAME_CPP
#define RACINGGAME_CPP
int counter = 0;
int row_x[10];
float fasttickdelay = 100; //75
boolean ison[15];
int fasttickercounter =0;
int gameend = 0;
int referenzpoint =4; //Point where the left roadside is
int pixelposx=referenzpoint+1;
int pixelposy=5;
boolean roadpieces[10][16];
unsigned long lasttick=0;
unsigned long lastfasttick=0;
boolean firstone = true;
int keepstate = 0; //Keep State meens a natural street


unsigned long millis() {
    struct timeval tp;
    gettimeofday(&tp, NULL);
    return (long long) tp.tv_sec * 1000L + tp.tv_usec / 1000; //get current timestamp in milliseconds
}

RacingGame::RacingGame() {

}


void reset(SystemInterface* systeminterface) {
    for(int i=0;i<10;i++){
        row_x[i]=0;
    }
    for(int i=0;i<15;i++){
        ison[i]=false;
    }

    for(int x=0;x<10;x++){
        for(int y=0;y<6;y++){
            roadpieces[x][y]=false;
        }
    }
    systeminterface->ledTisch->clear();
    systeminterface->ledTisch->show();

    for(int x=0;x<10;x++) {
        for(int y=0;y<15;y++) {
            roadpieces[x][y] = false;
        }

    }

    firstone = true;

    pixelposy = 5;
    pixelposx = 5;
    keepstate = 0;
    gameend = 0;
    fasttickdelay = 100;
    fasttickercounter =1;

}

void RacingGame::onCreate(SystemInterface* systeminterface){
    lasttick = millis();
    lastfasttick = millis();

    counter =0;


    reset(systeminterface);
}


void tick()
{



    if(keepstate==0) {
        if(random()%10>7) {
            if (random()%10 > 4)
            {
                keepstate++;

                if (referenzpoint != 0)
                {
                    referenzpoint--;
                }
            }
            else
            {
                keepstate++;

                if (referenzpoint < 5)
                {
                    referenzpoint++;
                }
            }

        }

    }else{
        keepstate++;
        if(keepstate==2) {
            keepstate = 0;
        }

    }




    if(firstone) {
        firstone = false;
        for(int i=0;i<referenzpoint+1;i++) {

            roadpieces[i][15] = true;

        }

        for(int i=referenzpoint+4;i<10;i++) {
            roadpieces[i][15] = true;
        }


    }



    for(int y=0;y<15;y++) {
        for(int x=0;x<10;x++) {

            roadpieces[x][y] = roadpieces[x][y+1];

        }


    }


    for(int i=0;i<10;i++) {
        roadpieces[i][15] = false;
    }


    roadpieces[referenzpoint][15] = true;
    roadpieces[referenzpoint+4][15] = true;

}

void RacingGame::onRun(SystemInterface* systemInterface) {

    if(gameend ==0) {


        if ((millis() - lastfasttick) >= fasttickdelay) {


            fasttickdelay -= 0.13f;

            tick();

            if (fasttickercounter == 3) {
                fasttickercounter = 0;
            }

            for (int i = 0; i < 15; i++) {
                ison[i] = false;

                if ((i + fasttickercounter) % 3 == 0) {
                    ison[i] = true;
                }
/*
    if(i==12+fasttickercounter) {

        ison[i] = true;

    }
*/

            }


            fasttickercounter++;
            lastfasttick = millis();

        }

    }


}

void RacingGame::onDraw(SystemInterface* systeminterface){

    if(gameend ==0) {


        if ((millis() - lastfasttick) >= fasttickdelay)
        {


            fasttickdelay-=0.13f;

            tick();

            if(fasttickercounter==3) {
                fasttickercounter=0;
            }

            for(int i=0;i<15;i++) {
                ison[i] = false;

                if((i+fasttickercounter)%3==0) {
                    ison[i] =true;
                }
/*
    if(i==12+fasttickercounter) {

        ison[i] = true;

    }
*/

            }



            fasttickercounter++;
            lastfasttick = millis();

        }


        int counter =0;
        for(int y=0;y<15;y++) {
            counter++;
            for(int x=0;x<10;x++) {



                if(roadpieces[x][y]) {

                    if(ison[y]) {



                        systeminterface->ledTisch->setcolor(0,160,0);
                        systeminterface->ledTisch->drawkoordinatensystem(x,y);

                    }else{


                        systeminterface->ledTisch->setcolor(100,100,100);
                        systeminterface->ledTisch->drawkoordinatensystem(x,y);



                    }

                }else{
                    systeminterface->ledTisch->setcolor(0,0,0);
                    systeminterface->ledTisch->drawkoordinatensystem(x,y);


                }

            }

            if(counter>2) {
                counter=0;
            }
        }





        systeminterface->ledTisch->setcolor(100,0,0);
        systeminterface->ledTisch->drawkoordinatensystem(pixelposx,pixelposy);
        systeminterface->ledTisch->drawkoordinatensystem(pixelposx,pixelposy+1);


        if(roadpieces[pixelposx][pixelposy]||roadpieces[pixelposx][pixelposy+1]) {
            for(int x=0;x<10;x++) {
                for(int y=0;y<15;y++) {
                    systeminterface->ledTisch->setcolor(100,0,0);
                    systeminterface->ledTisch->drawkoordinatensystem(x,y);

                }
            }

            gameend = 1;


        }
    }


}



void RacingGame::onDataReceive(char* data, SystemInterface* systeminterface,int ControllerID){

    if(strcmp(data,"h\n")==0){//hoch
        pixelposy++;
    }

    if(strcmp(data,"r\n")==0){//rechts
        pixelposx++;
    }

    if(strcmp(data,"l\n")==0 ){//links
        pixelposx--;
    }

    if(strcmp(data,"d\n")==0){//runter
        pixelposy--;
    }
    if(strcmp(data,"n\n")==0) {//new game
        reset(systeminterface);

    }

}
void RacingGame::onStop(SystemInterface* systeminterface){

}

char* RacingGame::getName(){
    return "RacingGame";
}

#endif

File: LEDWall2.0/src/main/java/de/ft/ledwall/apps/racingGame/RacingGame.kt
package de.ft.ledwall.apps.racingGame

import de.ft.ledwall.Application
import de.ft.ledwall.SystemInterface
import de.ft.ledwall.animation.AnimationManager


class RacingGame: Application {
    var counter:Int = 0
    var row_x = IntArray(10)
    var fasttickdelay:Float = 100f //75
    var ison = BooleanArray(15)
    var fasttickercounter:Int =0
    var gameend:Int = 0
    var referenzpoint:Int =4 //Point where the left roadside is
    var pixelposx:Int=referenzpoint+1
    var pixelposy:Int=5
    var roadpieces = Array(10) {BooleanArray(16)}
    var  lasttick:Long=0
    var lastfasttick:Long=0
    var firstone:Boolean = true
    var keepstate:Int = 0 //Keep State meens a natural street


    var ani_manager:AnimationManager = AnimationManager()

    fun reset() {
        for(i in 0 until 10){
            row_x[i]=0
        }
        for(i in 0 until 15){
            ison[i]=false
        }

        for(x in 0 until 10){
            for(y in 0 until 6){
            roadpieces[x][y]=false
        }
        }
        SystemInterface.table.clear()

        for(x in 0 until 10) {
            for(y in 0 until 15) {
            roadpieces[x][y] = false
        }

        }

        firstone = true

        pixelposy = 5
        pixelposx = 5
        keepstate = 0
        gameend = 0
        fasttickdelay = 100f
        fasttickercounter =1

    }

    private fun tick() {
        if (keepstate == 0) {
            if (Math.random()*Int.MAX_VALUE % 10 > 7) {
                if (Math.random()*Int.MAX_VALUE % 10 > 4) {
                    keepstate++
                    if (referenzpoint != 0) {
                        referenzpoint--
                    }
                } else {
                    keepstate++
                    if (referenzpoint < 5) {
                        referenzpoint++
                    }
                }
            }
        } else {
            keepstate++
            if (keepstate == 2) {
                keepstate = 0
            }
        }
        if (firstone) {
            firstone = false
            for (i in 0 until referenzpoint + 1) {
                roadpieces[i][15] = true
            }
            for (i in referenzpoint + 4..9) {
                roadpieces[i][15] = true
            }
        }
        for (y in 0..14) {
            for (x in 0..9) {
                roadpieces[x][y] = roadpieces[x][y + 1]
            }
        }
        for (i in 0..9) {
            roadpieces[i][15] = false
        }
        roadpieces[referenzpoint][15] = true
        roadpieces[referenzpoint + 4][15] = true
    }


    override fun onCreate() {


        lasttick = System.currentTimeMillis()
        lastfasttick = System.currentTimeMillis()

        ani_manager.addToQueue(AnimationManager.test)
        counter =0
        reset()
    }

    override fun onDraw() {
        if(ani_manager.update()) return

        if (gameend == 0) {
            var counter = 0
            for (y in 0..14) {
                counter++
                for (x in 0..9) {
                    if (roadpieces[x][y]) {
                        if (ison[y]) {

                            SystemInterface.table.setColor(0,160,0)
                            SystemInterface.table.drawPixel(x,y)
                        } else {
                            SystemInterface.table.setColor(100,100,100)
                            SystemInterface.table.drawPixel(x,y)
                        }
                    } else {
                        SystemInterface.table.setColor(0,0,0)
                        SystemInterface.table.drawPixel(x,y)
                    }
                }
                if (counter > 2) {
                    counter = 0
                }
            }
            SystemInterface.table.setColor(100,0,0)
            SystemInterface.table.drawPixel(pixelposx,pixelposy)
            SystemInterface.table.drawPixel(pixelposx,pixelposy+1)
            if (roadpieces[pixelposx][pixelposy] || roadpieces[pixelposx][pixelposy + 1]) {
                for (x in 0..9) {
                    for (y in 0..14) {
                        SystemInterface.table.setColor(100,0,0)
                        SystemInterface.table.drawPixel(x,y)
                    }
                }
                gameend = 1
            }
        }

    }

    override fun onRun() {
        if(ani_manager.animationsAvailable()) return
        if (gameend == 0) {
            if (System.currentTimeMillis() - lastfasttick >= fasttickdelay) {
                fasttickdelay -= 0.13f
                tick()
                if (fasttickercounter == 3) {
                    fasttickercounter = 0
                }
                for (i in 0..14) {
                    ison[i] = false
                    if ((i + fasttickercounter) % 3 == 0) {
                        ison[i] = true
                    }
                    /*
    if(i==12+fasttickercounter) {

        ison[i] = true

    }
*/
                }
                fasttickercounter++
                lastfasttick = System.currentTimeMillis()
            }
        }
    }

    override fun onDataReceive(data: String, playerID: Int) {
        if(data.contentEquals("h")) pixelposy++
        if(data.contentEquals("r")) pixelposx++
        if(data.contentEquals("l")) pixelposx--
        if(data.contentEquals("d")) pixelposy--

        if(data.contentEquals("n")) reset()
    }

    override fun getName(): String {
        return "RacingGame"
    }

    override fun onStop() {
       
    }


}
